File: templates/src/sqlite-query.ts
/**
 * 模板:SQLite 数据库 → ACP 元件(只读查询)
 *
 * 用 Node 22 内建的 node:sqlite(零依赖)把一个 SQLite 库暴露成 db.query 元件。
 * 内置只读防护:仅允许 SELECT / WITH 开头的单条语句,杜绝写操作与多语句注入。
 *
 * 启动:pnpm start:sqlite-query(或 node --experimental-strip-types src/sqlite-query.ts)
 */
import { createRequire } from "node:module";
import type Sqlite from "node:sqlite";
import { AcpServer, defineComponent } from "@xi-era/acp-sdk/server";

// node:sqlite 在 Node 22.5+ 可用;经 createRequire 引入以兼容打包/测试工具的解析器
const { DatabaseSync } = createRequire(import.meta.url)("node:sqlite") as typeof Sqlite;

// ★ 改这里:你的数据库文件(默认用内存库 + 示例表)
const DB_PATH = process.env["ACP_DB_PATH"] ?? ":memory:";
const db = new DatabaseSync(DB_PATH);
db.exec("CREATE TABLE IF NOT EXISTS books (id INTEGER PRIMARY KEY, title TEXT, year INTEGER)");
if (db.prepare("SELECT COUNT(*) AS n FROM books").get() !== undefined) {
  const { n } = db.prepare("SELECT COUNT(*) AS n FROM books").get() as { n: number };
  if (n === 0) {
    db.prepare("INSERT INTO books (title, year) VALUES (?, ?)").run("Designing Data-Intensive Applications", 2017);
    db.prepare("INSERT INTO books (title, year) VALUES (?, ?)").run("The Pragmatic Programmer", 1999);
  }
}

/** 只读白名单:拒绝写操作、多语句、PRAGMA 等危险前缀 */
function assertReadOnly(sql: string): void {
  const trimmed = sql.trim().replace(/;+\s*$/, "");
  if (/;/.test(trimmed)) throw new Error("multiple statements are not allowed");
  if (!/^(select|with)\b/i.test(trimmed)) throw new Error("only SELECT / WITH queries are allowed");
}

export const sqliteQuery = defineComponent({
  id: "db.query",
  name: "SQLite Query",
  description: "Runs a read-only SQL query against the template SQLite database",
  version: "1.0.0",
  inputSchema: { type: "object", properties: { sql: { type: "string" } }, required: ["sql"] },
  tags: ["db", "template"],
  async handle(input: { sql: string }) {
    assertReadOnly(input.sql);
    const rows = db.prepare(input.sql).all();
    return { rows, count: rows.length };
  },
});

const isMain =
  process.argv[1]?.replace(/\\/g, "/").endsWith("sqlite-query.ts") ||
  process.argv[1]?.replace(/\\/g, "/").endsWith("sqlite-query.js");
if (isMain) {
  const server = new AcpServer({ name: "sqlite-node", version: "1.0.0" });
  server.register(sqliteQuery);
  server
    .listen({ port: Number(process.env["ACP_PORT"] ?? 8092) })
    .then(({ port }) => console.log(`SQLite ACP endpoint: http://localhost:${port}/acp`));
}

File: templates/python/sqlite_query.py
"""模板:SQLite 数据库 → ACP 元件(只读查询,Python 版)。

用标准库 sqlite3(零依赖)。内置只读防护。
启动:python templates/python/sqlite_query.py
"""
from __future__ import annotations

import os
import sqlite3
from typing import Any

from acp import AcpServer
from acp.component import ComponentDef

DB_PATH = os.environ.get("ACP_DB_PATH", ":memory:")
db = sqlite3.connect(DB_PATH, check_same_thread=False)
db.execute("CREATE TABLE IF NOT EXISTS books (id INTEGER PRIMARY KEY, title TEXT, year INTEGER)")
db.execute("INSERT OR IGNORE INTO books (title, year) VALUES ('Designing Data-Intensive Applications', 2017)")
db.execute("INSERT OR IGNORE INTO books (title, year) VALUES ('The Pragmatic Programmer', 1999)")
db.commit()


def assert_read_only(sql: str) -> None:
    trimmed = sql.strip().rstrip(";").strip()
    if ";" in trimmed:
        raise ValueError("multiple statements are not allowed")
    if not trimmed.lower().startswith(("select", "with")):
        raise ValueError("only SELECT / WITH queries are allowed")


def handle_query(inp: Any, ctx) -> dict[str, Any]:  # noqa: ANN001
    sql = (inp or {}).get("sql", "")
    assert_read_only(sql)
    rows = [dict(zip([c[0] for c in cur.description], row)) for cur in [db.execute(sql)] for row in cur.fetchall()]
    return {"rows": rows, "count": len(rows)}


sqlite_query = ComponentDef(
    id="db.query",
    name="SQLite Query",
    description="Runs a read-only SQL query against the template SQLite database",
    input_schema={"type": "object", "properties": {"sql": {"type": "string"}}, "required": ["sql"]},
    tags=["db", "template"],
    handle=handle_query,
)

if __name__ == "__main__":
    server = AcpServer(name="sqlite-node", version="1.0.0")
    server.register(sqlite_query)
    ws_port = server.listen(port=int(os.environ.get("ACP_PORT", "8092")))
    print(f"HTTP ACP endpoint: http://localhost:{server.http_port}/acp  (WS: ws://localhost:{ws_port}/acp)")
    server.serve_forever()
